File: vreader/Models/VReaderLocator.swift
// Purpose: Durable, engine-agnostic reading-position envelope (Feature #42).
// Wraps an engine's locator representation behind a vreader-owned value type so
// saved positions survive an engine swap / re-conversion. Never persist a raw
// engine-internal anchor — persist this envelope instead.
//
// Key decisions:
// - Value type (Codable/Equatable/Sendable), NOT a @Model. Stored on
//   ReadingPosition as raw `Data?` (SchemaV8), mirroring Highlight.anchorData's
//   SwiftData-safe precedent — a Codable struct decoded with try? rather than a
//   @Model column SwiftData would have to materialize.
// - `engine` disambiguates which locator representation is authoritative during
//   the two-engine era: `.epubWKWebView` (legacy `Locator`) vs `.readium`
//   (`readiumLocatorJSON`, Readium's Locator serialized as plain JSON — so this
//   type needs NO Readium dependency).
// - `legacyLocator` carries the existing per-format `Locator` for back-compat /
//   dual-write, so both engines persist through one envelope.
// - `canonicalHash` mirrors Locator.canonicalHash / AnnotationAnchor.anchorHash:
//   sorted-keys JSON, SHA-256, hex string. Deterministic + stable across
//   encode/decode round-trips for dedup/sync keys.
//
// @coordinates-with: Locator.swift, ReadingPosition.swift, SchemaV8.swift

import Foundation
import CryptoKit

/// Which reader engine produced the authoritative locator in a VReaderLocator.
enum ReaderLocatorEngine: String, Codable, Hashable, Sendable, CaseIterable {
    /// The legacy bespoke EPUBWebViewBridge engine — locator lives in `legacyLocator`.
    case epubWKWebView
    /// The Readium Swift Toolkit engine — locator lives in `readiumLocatorJSON`.
    case readium
}

/// Durable engine-agnostic position envelope. Wraps an engine locator + survives
/// re-conversion. Persisted as JSON-encoded `Data?` on `ReadingPosition`.
struct VReaderLocator: Codable, Equatable, Sendable {
    /// The book's `DocumentFingerprint.canonicalKey`.
    let fingerprintKey: String

    /// The book's original (imported) format.
    let originalFormat: BookFormat

    /// Which engine's locator is authoritative.
    let engine: ReaderLocatorEngine

    /// Readium's `Locator` serialized as JSON (stored as a plain String so this
    /// type carries no Readium dependency). nil for the legacy engine.
    let readiumLocatorJSON: String?

    /// The existing per-format vreader `Locator`, for back-compat / dual-write.
    /// nil when only a Readium locator is available.
    let legacyLocator: Locator?

    /// Envelope schema version (lets future fields evolve independently of the
    /// SwiftData schema version).
    let schemaVersion: Int

    // MARK: - Init

    init(
        fingerprintKey: String,
        originalFormat: BookFormat,
        engine: ReaderLocatorEngine,
        readiumLocatorJSON: String?,
        legacyLocator: Locator?,
        schemaVersion: Int = VReaderLocator.currentSchemaVersion
    ) {
        self.fingerprintKey = fingerprintKey
        self.originalFormat = originalFormat
        self.engine = engine
        self.readiumLocatorJSON = readiumLocatorJSON
        self.legacyLocator = legacyLocator
        self.schemaVersion = schemaVersion
    }

    /// Current envelope schema version.
    static let currentSchemaVersion = 1

    // MARK: - Legacy wrapping

    /// Wraps an existing `Locator` as a legacy-engine envelope. Derives
    /// `fingerprintKey` and `originalFormat` from the locator's fingerprint, so
    /// both engines persist through one model.
    init(legacyLocator: Locator, schemaVersion: Int = VReaderLocator.currentSchemaVersion) {
        self.fingerprintKey = legacyLocator.bookFingerprint.canonicalKey
        self.originalFormat = legacyLocator.bookFingerprint.format
        self.engine = .epubWKWebView
        self.readiumLocatorJSON = nil
        self.legacyLocator = legacyLocator
        self.schemaVersion = schemaVersion
    }

    // MARK: - Canonical Hash

    /// SHA-256 hash of the canonical (sorted-keys) JSON encoding. Deterministic
    /// and stable across encode/decode round-trips. Used for dedup/sync keys.
    var canonicalHash: String {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys, .withoutEscapingSlashes]
        // VReaderLocator is always encodable (all fields are Codable value types).
        let data = (try? encoder.encode(self)) ?? Data()
        let digest = SHA256.hash(data: data)
        return digest.map { String(format: "%02x", $0) }.joined()
    }
}

File: android/identity/src/main/kotlin/vreader/contracts/VReaderLocator.kt
package vreader.contracts

import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.security.MessageDigest

/**
 * Which reader engine produced the authoritative locator in a [VReaderLocator]
 * (mirrors Swift `ReaderLocatorEngine`). Exactly two persisted cases.
 */
enum class ReaderLocatorEngine {
    /** The legacy bespoke EPUB WKWebView engine — locator lives in `legacyLocator`. */
    epubWKWebView,

    /** The Readium toolkit engine — locator lives in `readiumLocatorJSON`. */
    readium,
}

/**
 * Durable, engine-agnostic reading-position envelope — the Kotlin mirror of Swift
 * `VReaderLocator` (vreader/Models/VReaderLocator.swift). The persistence layer
 * stores THIS, never a bare [Locator] (feature #106 Gate-2 Critical): saved
 * positions survive an engine swap / re-conversion because the envelope records
 * which engine is authoritative plus the canonical fallback.
 *
 * - `readiumLocatorJSON` — Readium's own CFI-bearing JSON (platform-local; the
 *   cross-platform fallback is `legacyLocator`'s progression + textQuote).
 * - `legacyLocator` — the engine-neutral [Locator] (canonical resume anchor).
 * - `schemaVersion` — the envelope's own migration hook, independent of the Room
 *   database schema version.
 *
 * `@Serializable` so the Room layer can JSON-encode the whole envelope into one
 * column (the iOS analog persists it as `Data?` on `ReadingPosition`).
 */
@Serializable
data class VReaderLocator(
    val fingerprintKey: String,
    val originalFormat: BookFormat,
    val engine: ReaderLocatorEngine,
    val readiumLocatorJSON: String? = null,
    val legacyLocator: Locator? = null,
    val schemaVersion: Int = CURRENT_SCHEMA_VERSION,
) {
    /**
     * SHA-256 (hex) of this envelope's deterministic JSON encoding — stable across
     * encode/decode round-trips, for local dedup/sync keys (mirrors Swift
     * `VReaderLocator.canonicalHash`).
     *
     * NOTE: this is **locally** deterministic (kotlinx fixed field order). It is
     * NOT yet asserted byte-equal to Swift's `JSONEncoder(.sortedKeys)` output —
     * cross-platform *envelope*-vector parity is the deferred WI-2 conformance
     * extension. Use it for on-device dedup, not (yet) for cross-device sync keys.
     */
    val canonicalHash: String
        get() {
            val bytes = CANONICAL_JSON.encodeToString(this).toByteArray(Charsets.UTF_8)
            return MessageDigest.getInstance("SHA-256").digest(bytes)
                .joinToString("") { "%02x".format(it) }
        }

    companion object {
        /** Current envelope schema version (independent of the Room DB version). */
        const val CURRENT_SCHEMA_VERSION = 1

        private val CANONICAL_JSON = Json { encodeDefaults = true }

        /**
         * Wraps an engine-neutral [Locator] as a legacy-engine envelope, deriving
         * `fingerprintKey`/`originalFormat` from the locator's identity triple —
         * mirrors Swift `init(legacyLocator:)`. The format string must be a valid
         * [BookFormat] raw value (it came from a `DocumentFingerprint`).
         */
        fun wrapLegacy(
            legacyLocator: Locator,
            schemaVersion: Int = CURRENT_SCHEMA_VERSION,
        ): VReaderLocator {
            val format = BookFormat.entries.firstOrNull { it.name == legacyLocator.format }
                ?: error("invalid BookFormat raw value: ${legacyLocator.format}")
            return VReaderLocator(
                fingerprintKey = legacyLocator.fingerprintKey,
                originalFormat = format,
                engine = ReaderLocatorEngine.epubWKWebView,
                readiumLocatorJSON = null,
                legacyLocator = legacyLocator,
                schemaVersion = schemaVersion,
            )
        }
    }
}
